market_maker_keeper/uniswapv2_market_maker_keeper.py

```diff
@@ -0,0 +1,379 @@
+            lifecycle.every(10, self.place_liquidity)
+
+    def startup(self):
+        self.uniswap.approve(self.token_a)
+        self.uniswap.approve(self.token_b)
+
+    def get_token_config(self):
+        current_config = self.reloadable_config.get_token_config()
+        if current_config != self._last_config_dict:
+            self._last_config = TokenConfig(current_config)
+            self._last_config_dict = current_config
+
+            self.logger.info(f"Successfully parsed configuration")
+
+        return self._last_config
+
+    def pair(self) -> str:
+        return self.arguments.pair
+
+    def calculate_liquidity_args(self, token_a_balance: Wad, token_b_balance: Wad) -> Optional[dict]:
+        """ Returns dictionary containing arguments for addLiquidity transactions
+
+        Calculate amount of both tokens, given the current reserve ratio on uniswap
+        Use accepted_slippage to calculate min off of available balance + liquidity
+
+        If eth is in the pair, at least 1 eth should be left for gas
+        """
+
+        if self.is_eth:
+            token_a_balance = token_a_balance - Wad.from_number(1)
+            if token_a_balance < Wad.from_number(0):
+                self.logger.info(f"Insufficient Eth balance.")
+                return
+
+        token_a_desired = min(token_a_balance, token_b_balance / self.uniswap_current_exchange_price)
+        token_a_min = token_a_desired - (token_a_desired * self.accepted_slippage)
+        token_b_desired = min(token_b_balance, token_a_desired * self.uniswap_current_exchange_price)
+        token_b_min = token_b_desired - (token_b_desired * self.accepted_slippage)
+
+        add_liquidity_args = {
+            'amount_a_desired': self.token_a.unnormalize_amount(token_a_desired),
+            'amount_b_desired': self.token_b.unnormalize_amount(token_b_desired),
+            'amount_a_min': self.token_a.unnormalize_amount(token_a_min),
+            'amount_b_min': self.token_b.unnormalize_amount(token_b_min)
+        }
+        return add_liquidity_args
+
+    def add_liquidity(self) -> Optional[Receipt]:
+        """ Send an addLiquidity or addLiquidityETH transaction to the UniswapV2 Router Contract.
+
+        UniswapV2 differentiates between ETH and ERC20 token transactions.
+        All operations are handled in WETH, but the Router Contract handles all wrapping and unwrapping.
+
+        It is assumed that all availability should be added, except for 1 Eth for future Gas operations.
+
+        Given available token balances and an exchange rate, calculations of amounts to add,
+        and limits for price movement need to be calculated.
+
+        Args:
+            uniswap_current_exchange_price: The current Product of the pool
+        Returns:
+            A Pymaker Receipt object or a None
+        """
+
+        token_a_balance = self.uniswap.get_account_token_balance(self.token_a) if not self.is_eth else self.uniswap.get_account_eth_balance()
```

Comment:
Yes, that was old code that has since been removed. I realized logic could be significantly simpler if Eth is assumed to be the base token. With this assumption, I can avoid a ton of logical complexity each time I query to check if the token is eth or not.

Reply:
That assumption has been removed, and ETH can now be either the base, or the quote currency.